packages/Kokkos/src/DTK_UserDataInterface.hpp

```diff
@@ -0,0 +1,273 @@
+//---------------------------------------------------------------------------//
+/*
+  Copyright (c) 2012, Stuart R. Slattery
+  All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are
+  met:
+
+  *: Redistributions of source code must retain the above copyright
+  notice, this list of conditions and the following disclaimer.
+
+  *: Redistributions in binary form must reproduce the above copyright
+  notice, this list of conditions and the following disclaimer in the
+  documentation and/or other materials provided with the distribution.
+
+  *: Neither the name of the University of Wisconsin - Madison nor the
+  names of its contributors may be used to endorse or promote products
+  derived from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+//---------------------------------------------------------------------------//
+/*!
+ * \file DTK_UserDataInterface.hpp
+ * \brief C++ DTK interface for accessing application data.
+ */
+//---------------------------------------------------------------------------//
+
+#ifndef DTK_USERDATAINTERFACE_HPP
+#define DTK_USERDATAINTERFACE_HPP
+
+#include "DTK_ConfigDefs.hpp"
+#include "DTK_View.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace DataTransferKit
+{
+//---------------------------------------------------------------------------//
+/*!
+ * \namespace UserDataInterface
+ *
+ * The type alias definitions here serve as argument templates for callable
+ * objects to be impelmented in the user's application for access to
+ * application data by DTK algorithms. The callable objects are meant to be
+ * called from the host.
+ *
+ * The first argument of each callable object is a void shared pointer to user
```

Comment:
We can suggest this through examples but I'm ok letting them do whatever they want. They know we are taking a shared_ptr and storing it somewhere